File: Data Structures & Algorithms/combination-target-sum-ii/submission-2.py
class Solution:
    def combinationSum2(self, nums: List[int], target: int) -> List[List[int]]:
        res = []

        def f(idx, curr, target):
            if target == 0:
                res.append(curr[:])
                return

            if target < 0 or idx == len(nums):
                return
            
            

            curr.append(nums[idx])
            f(idx + 1, curr, target - nums[idx])
            curr.pop()

            for i in range(idx+1, len(nums)):
                if nums[i] != nums[idx]:
                    f(i, curr, target)
                    break

           
        nums.sort()
        f(0, [], target)

        return res

File: Data Structures & Algorithms/combination-target-sum/submission-2.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> curr = new ArrayList<>();

        generateSum(0, nums, target, curr, res);
        return res;
    }

    private void generateSum(int idx, int[] nums, int target, List<Integer> curr, List<List<Integer>> res) {
        if(target == 0) {
            res.add(new ArrayList<>(curr));
            return;
        }
        
        if(target < 0 || idx == nums.length) return;

        curr.add(nums[idx]);
        generateSum(idx, nums, target - nums[idx], curr, res);

        curr.remove(curr.size()-1);
        generateSum(idx+1, nums, target, curr, res);
        
    }
}
